src/System.Threading.Timer/tests/TimerFiringTests.cs

```diff
@@ -2,339 +2,361 @@
+            AutoResetEvent are = new AutoResetEvent(false);
 
-    [Fact]
-    public void Timer_CanBeDisposedMultipleTimes()
-    {
-        // There's nothing to validate besides that we don't throw an exception, so rely on xunit 
-        // to catch any exception that would be thrown and signal a test failure
-        TimerCallback tc = new TimerCallback((object o) => { });
-        var t = new Timer(tc, null, 100, -1);
-        for (int i = 0; i < 10; i++)
-            t.Dispose();
-    }
+            using (var t = new Timer(new TimerCallback((object s) =>
+            {
+                if (Interlocked.Increment(ref count) >= 2)
+                {
+                    are.Set();
+                }
+            }), null, TimeSpan.FromMilliseconds(dueTime), TimeSpan.FromMilliseconds(period)))
+            {
+                Assert.True(are.WaitOne(TimeSpan.FromMilliseconds(MaxPositiveTimeoutInMs)));
+            }
+        }
 
-    [Fact]
-    public void NonRepeatingTimer_ThatHasAlreadyFired_CanChangeAndFireAgain()
-    {
-        AutoResetEvent are = new AutoResetEvent(false);
-        TimerCallback tc = new TimerCallback((object o) => are.Set());
-        using (var t = new Timer(tc, null, 1, Timeout.Infinite))
+        [Fact]
+        public void Timer_FiresOnlyOnce_OnDueTime_With_InfinitePeriod()
         {
-            Assert.True(are.WaitOne(MaxPositiveTimeoutInMs), "Should have received first timer event");
-            Assert.False(are.WaitOne(500), "Should not have received a second timer event");
-            t.Change(10, Timeout.Infinite);
-            Assert.True(are.WaitOne(MaxPositiveTimeoutInMs), "Should have received a second timer event after changing it");
+            int count = 0;
+            AutoResetEvent are = new AutoResetEvent(false);
+
+            using (var t = new Timer(new TimerCallback((object s) =>
+            {
+                if (Interlocked.Increment(ref count) >= 2)
+                {
+                    are.Set();
+                }
+            }), null, TimeSpan.FromMilliseconds(100), TimeSpan.FromMilliseconds(Timeout.Infinite) /* not relevant */))
+            {
+                Assert.False(are.WaitOne(TimeSpan.FromMilliseconds(250 /*enough for 2 fires + buffer*/)));
+            }
         }
-    }
 
-    [Fact]
-    public void MultpleTimers_PeriodicTimerIsntBlockedByBlockedCallback()
-    {
-        int callbacks = 2;
-        Barrier b = new Barrier(callbacks + 1);
-        Timer t = null;
-        t = new Timer(_ =>
+        [OuterLoop("Waits for a second")]
+        [Fact]
+        public void Timer_ChangeToDelete_DoesntFire()
         {
-            if (Interlocked.Decrement(ref callbacks) >= 0)
+            const int DueTime = 1000;
+            var are = new AutoResetEvent(false);
+            using (var t = new Timer(_ => are.Set(), null, DueTime, -1))
             {
-                Assert.True(b.SignalAndWait(MaxPositiveTimeoutInMs));
+                t.Change(-1, -1);
```

Comment:
Nit, `Timeout.Infinite` is more readable than `-1` (would have saved me looking at MSDN)